Source/reactions/Castro_react_util.H

```diff
@@ -0,0 +1,265 @@
+#include <burn_type.H>
+// #include <actual_rhs.H>
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void
+single_zone_react_source(GpuArray<Real, NUM_STATE> const& state,
+                         GpuArray<Real, NUM_STATE>& R,
+                         burn_t burn_state);
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void 
+numerical_jac(burn_t& state, RArray2D& jac) {
+    return;
+}
+
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+bool
+okay_to_burn(const int i, const int j, const int k,
+             Array4<const Real> const& state) {
+
+    if (state(i,j,k,UTEMP) < react_T_min || state(i,j,k,UTEMP) > react_T_max ||
+        state(i,j,k,URHO) < react_rho_min || state(i,j,k,URHO) > react_rho_max) {
+            return false;
+    }
+
+    return true;
+}
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+bool
+okay_to_burn(GpuArray<Real, NUM_STATE> const& state) {
+
+    if (state[UTEMP] < react_T_min || state[UTEMP] > react_T_max ||
+        state[URHO] < react_rho_min || state[URHO] > react_rho_max) {
+        return false;
+    }
+
+    return true;
+}
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+bool
+okay_to_burn_type(burn_t const& state) {
+
+    if (state.T < react_T_min || state.T > react_T_max ||
+        state.rho < react_rho_min || state.rho > react_rho_max) {
+            return false;
+    }
+
+    return true;
+}
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void
+single_zone_react_source(const int i, const int j, const int k,
+                         Array4<const Real> const& state,
+                         Array4<Real> const& R,
+                         burn_t burn_state) {
+    GpuArray<Real, NUM_STATE> state_arr;
+    GpuArray<Real, NUM_STATE> R_arr; 
+
+    for (int n = 0; n < NUM_STATE; ++n) {
+        state_arr[n] = state(i,j,k,n);
+        R_arr[n] = R(i,j,k,n);
+    }
+
+    single_zone_react_source(state_arr, R_arr, burn_state);
+
+    for (int n = 0; n < NUM_STATE; ++n) {
+        R(i,j,k,n) = R_arr[n];
+    }
+}
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void
+single_zone_react_source(GpuArray<Real, NUM_STATE> const& state,
+                         GpuArray<Real, NUM_STATE>& R,
+                         burn_t burn_state) {
+
+    auto rhoInv = 1.0_rt / state[URHO];
+
+    burn_state.rho = state[URHO];
+    burn_state.T = state[UTEMP];
+    burn_state.e = state[UEINT] * rhoInv;
+
+    for (int n = 0; n < NumSpec; ++n) {
+       burn_state.xn[n] = amrex::max(amrex::min(state[UFS+n] * rhoInv, 1.0_rt), small_x);
+    }
+
+#if NAUX_NET > 0
+    for (int n = 0; n < NumAux; ++n) {
+       burn_state.aux[n] = state[UFX+n] * rhoInv;
+    }
+#endif
+
+    eos_t eos_state;
+
+    // Ensure that the temperature going in is consistent with the internal energy.
+    burn_to_eos(burn_state, eos_state);
+    eos(eos_input_re, eos_state);
+    eos_to_burn(eos_state, burn_state);
+
+    // eos_get_small_temp(&small_temp);
+    burn_state.T = amrex::min(MAX_TEMP, amrex::max(burn_state.T, small_temp));
+
+    burn_state.self_heat = false;
+
+    Array1D<Real, 1, VODE_NEQS> ydot;
+
```

Comment:
I think this should be dimensioned as `neqs` not `VODE_NEQS`, since `neqs` is what is used in the actual burners